tests/bot/utils/test_time.py

```diff
@@ -0,0 +1,87 @@
+        )
+
+        for delta, precision, max_units, expected in test_cases:
+            self.assertEqual(time.humanize_delta(delta, precision, max_units), expected)
+
+    def test_humanize_delta_raises_for_invalid_max_units(self):
+        test_cases = (-1, 0)
+
+        for max_units in test_cases:
+            with self.assertRaises(ValueError) as error:
+                time.humanize_delta(relativedelta(days=2, hours=2), 'hours', max_units)
+                self.assertEqual(str(error), 'max_units must be positive')
+
+    def test_parse_rfc1123(self):
+        """Testing parse_rfc1123."""
+        test_cases = (
+            ('Sun, 15 Sep 2019 12:00:00 GMT', datetime(2019, 9, 15, 12, 0, 0, tzinfo=timezone.utc)),
+        )
+
+        for stamp, expected in test_cases:
+            self.assertEqual(time.parse_rfc1123(stamp), expected)
+
+    @patch('asyncio.sleep', new_callable=AsyncMock)
+    def test_wait_until(self, mock):
+        """Testing wait_until."""
+        start = datetime(2019, 1, 1, 0, 0)
+        then = datetime(2019, 1, 1, 0, 10)
+
+        # No return value
+        assert asyncio.run(time.wait_until(then, start)) is None
+
+        mock.assert_called_once_with(10 * 60)
+
+    def test_format_infraction_with_duration(self):
+        """Testing format_infraction_with_duration."""
+        test_cases = (
+            ('2019-12-12T00:01:00Z', datetime(2019, 12, 11, 12, 0, 5), 2, '2019-12-12 00:01 (12 hours and 55 seconds)'),
+            ('2019-12-12T00:01:00Z', datetime(2019, 12, 11, 12, 0, 5), 1, '2019-12-12 00:01 (12 hours)'),
+            ('2019-12-12T00:01:00Z', datetime(2019, 12, 11, 12, 5, 5), 6,
+             '2019-12-12 00:01 (11 hours, 55 minutes and 55 seconds)'),
+            ('2019-12-12T00:00:00Z', datetime(2019, 12, 11, 23, 59), 2, '2019-12-12 00:00 (1 minute)'),
+            ('2019-11-23T20:09:00Z', datetime(2019, 11, 15, 20, 15), 2, '2019-11-23 20:09 (7 days and 23 hours)'),
+            ('2019-11-23T20:09:00Z', datetime(2019, 4, 25, 20, 15), 2, '2019-11-23 20:09 (6 months and 28 days)'),
+            ('2019-11-23T20:09:00Z', datetime(2019, 4, 25, 20, 15), 6,
+             '2019-11-23 20:09 (6 months, 28 days, 23 hours and 54 minutes)'),
+            ('2019-11-23T20:58:00Z', datetime(2019, 11, 23, 20, 53), 2, '2019-11-23 20:58 (5 minutes)'),
+            ('2019-11-24T00:00:00Z', datetime(2019, 11, 23, 23, 59, 0), 2, '2019-11-24 00:00 (1 minute)'),
+            ('2019-11-23T23:59:00Z', datetime(2017, 7, 21, 23, 0), 2, '2019-11-23 23:59 (2 years and 4 months)'),
+            ('2019-11-23T23:59:00Z', datetime(2019, 11, 23, 23, 49, 5), 2,
+             '2019-11-23 23:59 (9 minutes and 55 seconds)'),
+            (None, datetime(2019, 11, 23, 23, 49, 5), 2, None),
+        )
+
+        for expiry, date_from, max_units, expected in test_cases:
+            self.assertEqual(time.format_infraction_with_duration(expiry, date_from, max_units), expected)
```

Comment:
These should be in a subtest otherwise this whole loop is treated as a single test case.

Reply:
This applies to the other tests as well

Reply:
Hmm, can you explain this a bit more? When I was testing, it will actually stop at the failed case when looping/

Do you mean splitting it into cases like:
- No max_units
- Custom max_units
- `None` expiry

And then put the cases there?

Reply:
Ves wrote a bit about it https://github.com/python-discord/bot/tree/master/tests#using-selfsubtest-for-independent-subtests

Reply:
Thank you!

Reply:
I've split them into proper independent tests, as well as added the test for `format_infraction`

Reply:
They still aren't using the subtest context manager.

Reply:
Ah, I see the `self.subTest` now, I will refactor them later tomorrow. Thank you for the swift reply!

Reply:
Actually, I might as well do it now before I forget. Does this look better?